Feature/gitlab ci integration

radmen opened the pull request:
## About this PR:

Full integration with Gitlab CI process and Docker images build/deployment

Changes contain updated `.gitlab-ci.yml` config with support of:
* test stage
* complete docker images build process (including pushing to private registry) - separated to two threads
* deploy jobs
* tagging stable images

Vanilla will use our base Docker PHP image (`desmart/php:7.0-fpm-alpine`).

### Environments:

![zrzut ekranu 2016-11-10 o 19 27 30](https://cloud.githubusercontent.com/assets/1190255/20189110/072740da-a77c-11e6-918b-ef2362175638.png)

Gitlab will support deploying two different environments:
* `test` on every change on `develop` branch
* `beta` on every change on `release/*` branch

URL to deployed version will be: `ENV.PROJECT_NAME.staging.desmart.com`

Each env is tagged and contains two jobs which can be run manualy: 
* db seed - will reset database, migrate it again and seed
* stop - remove given environment from our staging server

## Questions:

* ~build process for `master` and tagged (eg `1.0.0`) version is basically the same (test, build images) IMO we can run them only for tagged versions. What do you think about it?~

## TODO:

- [x] Gitlab integration docs

Commit:
Copy gitlabci and docker config from private sandbox repo

Commit:
Ignore vim swp files

Commit:
Add dockerignore

Commit:
Include assets compilation during image build

Commit:
Fix ci config for building stable images

radmen commented on docker/php7/Dockerfile:
```diff
@@ -1,24 +1,24 @@
+    && php artisan optimize \
+    && php artisan vendor:publish
+
+VOLUME /usr/src/app/public/upload
```
this should be removed

fojuth commented:
TBH, I have to little knowledge about Docker to discuss all of this. If you think something should be done - do it. I'me relying on your knowledge and experience. Godspeed :)

radmen commented on .gitlab-ci.yml:
```diff
@@ -1,15 +1,169 @@
+        - export COMPOSE_FILE=docker-compose.local.yml
+        - docker-compose pull
+        - docker-compose build
+        - docker-compose run -e APP_KEY=$APP_KEY php make test
```
Add:

``` yml
after_script:
    - *COMPOSE_PROJECT_NAME
    - docker-compose stop
    - docker-compose rm --force
```

radmen commented on .gitlab-ci.yml:
```diff
@@ -1,15 +1,169 @@
+        - docker push $WWW_IMAGE
+    only:
+        - master
+        - develop
```
missing 

```
- *TAG_PATTERN
```

Commit:
Fixes for MR

Commit:
Add VOLUME for /usr/src/app/public/upload

Since vanilla creates (and possibly uses) this dir it makes sense to
mark it as a Docker VOLUME.

Commit:
Update README file

Commit:
Merge branch 'develop' into feature/gitlab-ci-integration

Conflicts:
  readme.md

radmen commented:
@fojuth it's no longer a WIP. Pls review it and if there's something that does not make sense for you let me know. 

I've updated also readme file. Pls check if it make sense.

radmen committed:
Remove VOLUME from PHP dockerfile

After some consideration I've decided that this is optional, and if app will support files upload the developer should have to manually set up this entry.

radmen committed:
Update readme.md
